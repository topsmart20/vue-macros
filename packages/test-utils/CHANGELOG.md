# @vue-macros/test-utils

## 1.0.1

### Patch Changes

- [`d944bb0`](https://github.com/sxzz/unplugin-vue-macros/commit/d944bb0e57bbf3a742f72a4776d582f52410a71f) Thanks [@sxzz](https://github.com/sxzz)! - tidy virtual module id

## 1.0.0

### Major Changes

- [`0b87d20`](https://github.com/sxzz/unplugin-vue-macros/commit/0b87d20f2581286421ad36fc5bc0768b0a2a44d0) Thanks [@sxzz](https://github.com/sxzz)! - Publish `@vue-macros/test-utils`

## 0.12.0
