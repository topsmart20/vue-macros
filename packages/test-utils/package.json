{
  "name": "@vue-macros/test-utils",
  "version": "1.1.3",
  "packageManager": "pnpm@8.7.6",
  "description": "test-utils feature from Vue Macros.",
  "type": "module",
  "keywords": [
    "vue-macros",
    "macros",
    "vue",
    "sfc",
    "setup",
    "script-setup",
    "test-utils"
  ],
  "license": "MIT",
  "homepage": "https://github.com/vue-macros/vue-macros#readme",
  "bugs": {
    "url": "https://github.com/vue-macros/vue-macros/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vue-macros/vue-macros.git",
    "directory": "packages/test-utils"
  },
  "author": "三咲智子 <sxzz@sxzz.moe>",
  "files": [
    "dist"
  ],
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "dev": "./src/index.ts",
      "require": "./dist/index.cjs",
      "import": "./dist/index.js"
    },
    "./*": [
      "./*",
      "./*.d.ts"
    ]
  },
  "scripts": {
    "build": "tsup",
    "dev": "tsup --watch"
  },
  "peerDependencies": {
    "vitest": "^0.29.0"
  },
  "peerDependenciesMeta": {
    "vitest": {
      "optional": true
    }
  },
  "dependencies": {
    "@rollup/plugin-node-resolve": "^15.2.1",
    "@rollup/pluginutils": "^5.0.4",
    "@vitejs/plugin-vue-jsx": "^3.0.2",
    "@vitejs/plugin-vue2": "^2.2.0",
    "esbuild": "^0.18.20",
    "fast-glob": "^3.3.1",
    "rollup": "^3.29.2",
    "rollup-plugin-esbuild": "^5.0.0",
    "unplugin-vue": "^4.3.5"
  },
  "devDependencies": {
    "@rollup/plugin-json": "^6.0.0",
    "vite": "^4.4.9",
    "vue": "^3.3.4"
  },
  "engines": {
    "node": ">=16.14.0"
  }
}
