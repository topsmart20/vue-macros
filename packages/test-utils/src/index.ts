export * from './rollup'
export * from './vitest'
export * from './strings'
