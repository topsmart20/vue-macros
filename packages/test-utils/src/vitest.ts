import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
