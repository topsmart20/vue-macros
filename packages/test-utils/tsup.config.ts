import { defineConfig } from 'tsup'
import cfg from '../../tsup.config.js'

export default defineConfig({
  ...cfg,
  entry: ['./src/index.ts'],
  splitting: false,
})
