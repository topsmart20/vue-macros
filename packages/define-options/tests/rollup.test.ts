import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import glob from 'fast-glob'
import { RollupToStringPlugin, rollupBuild } from '@vue-macros/test-utils'
import VueDefineOptions from '../src/rollup'

describe('Rollup', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const code = await rollupBuild(filepath, [
          VueDefineOptions({}),
          RollupToStringPlugin(),
        ]).catch((err) => err)
        expect(code).toMatchSnapshot()
      })
    }
  })
})
