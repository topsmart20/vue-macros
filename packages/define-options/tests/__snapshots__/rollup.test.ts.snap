// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/basic.vue 1`] = `
"var basic = \`<script>
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'Foo',
});
</script>
<script setup>

</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/el-button.vue 1`] = `
"var elButton = \`<script>
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'ElButton',
});
</script>
<template>
  <button
    ref="buttonRef"
    :class="[
      ns.b(),
      ns.m(buttonType),
      ns.m(buttonSize),
      ns.is('disabled', buttonDisabled),
      ns.is('loading', loading),
      ns.is('plain', plain),
      ns.is('round', round),
      ns.is('circle', circle),
    ]"
    :disabled="buttonDisabled || loading"
    :autofocus="autofocus"
    :type="nativeType"
    :style="buttonStyle"
    @click="handleClick"
  >
    <template v-if="loading">
      <slot v-if="$slots.loading" name="loading"></slot>
      <el-icon v-else :class="ns.is('loading')">
        <component :is="loadingIcon" />
      </el-icon>
    </template>
    <el-icon v-else-if="icon">
      <component :is="icon" />
    </el-icon>
    <span
      v-if="$slots.default"
      :class="{ [ns.em('text', 'expand')]: shouldAddSpace }"
    >
      <slot></slot>
    </span>
  </button>
</template>

<script setup>
import { computed, inject, Text, ref, useSlots } from 'vue'
import { useCssVar } from '@vueuse/core'
import { TinyColor } from '@ctrl/tinycolor'
import { ElIcon } from '@element-plus/components/icon'
import {
  useDisabled,
  useFormItem,
  useGlobalConfig,
  useNamespace,
  useSize,
} from '@element-plus/hooks'
import { buttonGroupContextKey } from '@element-plus/tokens'
import { buttonEmits, buttonProps } from './button'



const props = defineProps(buttonProps)
const emit = defineEmits(buttonEmits)
const slots = useSlots()

const buttonRef = ref()
const buttonGroupContext = inject(buttonGroupContextKey, undefined)
const globalConfig = useGlobalConfig('button')
const ns = useNamespace('button')
const autoInsertSpace = computed(
  () => props.autoInsertSpace ?? globalConfig.value?.autoInsertSpace ?? false,
)

// add space between two characters in Chinese
const shouldAddSpace = computed(() => {
  const defaultSlot = slots.default?.()
  if (autoInsertSpace.value && defaultSlot?.length === 1) {
    const slot = defaultSlot[0]
    if (slot?.type === Text) {
      const text = slot.children
      return /^\\p{Unified_Ideograph}{2}$/u.test(text)
    }
  }
  return false
})

const { form } = useFormItem()
const buttonSize = useSize(computed(() => buttonGroupContext?.size))
const buttonDisabled = useDisabled()
const buttonType = computed(() => props.type || buttonGroupContext?.type || '')

// calculate hover & active color by color
const typeColor = computed(() => useCssVar(\\\`--el-color-\${props.type}\\\`).value)
const buttonStyle = computed(() => {
  let styles = {}

  const buttonColor = props.color || typeColor.value

  if (buttonColor) {
    const shadeBgColor = new TinyColor(buttonColor).shade(10).toString()
    if (props.plain) {
      styles = {
        '--el-button-bg-color': new TinyColor(buttonColor).tint(90).toString(),
        '--el-button-text-color': buttonColor,
        '--el-button-hover-text-color': 'var(--el-color-white)',
        '--el-button-hover-bg-color': buttonColor,
        '--el-button-hover-border-color': buttonColor,
        '--el-button-active-bg-color': shadeBgColor,
        '--el-button-active-text-color': 'var(--el-color-white)',
        '--el-button-active-border-color': shadeBgColor,
      }
    } else {
      const tintBgColor = new TinyColor(buttonColor).tint(20).toString()
      styles = {
        '--el-button-bg-color': buttonColor,
        '--el-button-border-color': buttonColor,
        '--el-button-hover-bg-color': tintBgColor,
        '--el-button-hover-border-color': tintBgColor,
        '--el-button-active-bg-color': shadeBgColor,
        '--el-button-active-border-color': shadeBgColor,
      }
    }

    if (buttonDisabled.value) {
      const disabledButtonColor = new TinyColor(buttonColor).tint(50).toString()
      styles['--el-button-disabled-bg-color'] = disabledButtonColor
      styles['--el-button-disabled-border-color'] = disabledButtonColor
    }
  }

  return styles
})

const handleClick = (evt) => {
  if (props.nativeType === 'reset') {
    form?.resetFields()
  }
  emit('click', evt)
}
</script>
\`;

export { elButton as default };
"
`;

exports[`fixtures > tests/fixtures/empty.vue 1`] = `
"var empty = \`<script setup>


console.log(123)
</script>
\`;

export { empty as default };
"
`;

exports[`fixtures > tests/fixtures/error1.vue 1`] = `[SyntaxError: \`defineOptions()\` in <script setup> cannot reference locally declared variables (name) because it will be hoisted outside of the setup() function.]`;

exports[`fixtures > tests/fixtures/error2.vue 1`] = `[SyntaxError: \`defineOptions()\` in <script setup> cannot reference locally declared variables (foo) because it will be hoisted outside of the setup() function.]`;

exports[`fixtures > tests/fixtures/error3.vue 1`] = `[SyntaxError: defineOptions cannot be used with default export within <script>.]`;

exports[`fixtures > tests/fixtures/error4.vue 1`] = `[SyntaxError: defineOptions() please use defineProps, defineEmits, defineExpose, or defineSlots instead.]`;

exports[`fixtures > tests/fixtures/error5.vue 1`] = `[SyntaxError: defineOptions() please use defineProps, defineEmits, defineExpose, or defineSlots instead.]`;

exports[`fixtures > tests/fixtures/error6.vue 1`] = `[SyntaxError: duplicate defineOptions() call]`;

exports[`fixtures > tests/fixtures/function.vue 1`] = `
"var _function = \`<script lang="ts">
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  foo() {
    console.log('hello')
  },
  bar: () => console.log('world'),
  baz: class {
    a() {
      const a = () => 'b'
      return a
    }
  },
  fun: function () {
    console.log('fun')
  },
  iife: (() => Math.random())(),
});
</script>
<script lang="ts" setup>
const a = 'hello'

</script>
\`;

export { _function as default };
"
`;

exports[`fixtures > tests/fixtures/imports.vue 1`] = `
"var imports = \`<script lang="ts">
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  components: {
    Basic,
  },
});
</script>
<script setup lang="ts">
import Basic from './basic.vue'


</script>
\`;

export { imports as default };
"
`;

exports[`fixtures > tests/fixtures/js.vue 1`] = `
"var js = \`<script>
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'Foo',
});
</script>
<script setup>


console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { js as default };
"
`;

exports[`fixtures > tests/fixtures/jsx.vue 1`] = `
"var jsx = \`<script>
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'Jsx',
});
</script>
<script setup>

console.log(<h1>jsx</h1>)
</script>
\`;

export { jsx as default };
"
`;

exports[`fixtures > tests/fixtures/no-transform.vue 1`] = `
"var noTransform = \`<script setup>
console.log('setup')
</script>

<script>
export default {
  name: 'Hello',
}
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { noTransform as default };
"
`;

exports[`fixtures > tests/fixtures/no-transform2.vue 1`] = `
"var noTransform2 = \`<script setup>
console.log('setup')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { noTransform2 as default };
"
`;

exports[`fixtures > tests/fixtures/reference-var.vue 1`] = `
"var referenceVar = \`<script>
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent(options);
</script>
<script setup>
import options from './options'

</script>
\`;

export { referenceVar as default };
"
`;

exports[`fixtures > tests/fixtures/typescript.vue 1`] = `
"var typescript = \`<script lang="ts">
import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'Foo',
});
</script>
<script lang="ts" setup>


console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript as default };
"
`;

exports[`fixtures > tests/fixtures/with-script.vue 1`] = `
"var withScript = \`<script>
const name = '123'

import { defineComponent as __MACROS_defineComponent } from "vue";
export default /*#__PURE__*/ __MACROS_defineComponent({
  name: 'Foo',
});</script>

<script setup>

</script>
\`;

export { withScript as default };
"
`;
